protocols/bizcoin/bizcoin.go

```diff
@@ -0,0 +1,819 @@
+		// append response
+		resp, err := bz.prepare.Response(bz.tempPrepareResponse)
+		bz.tprMut.Unlock()
+		if err != nil {
+			return err
+		}
+		bzrReturn.Response = resp
+	} else {
+		bz.tprMut.Unlock()
+	}
+
+	dbg.Lvl3("BizCoin Handle Response PREPARE")
+	// if I'm root, we are finished, let's notify the "commit" round
+	if bz.IsRoot() {
+		// notify listeners (simulation) we finished
+		if bz.onResponsePrepareDone != nil {
+			bz.onResponsePrepareDone()
+		}
+		bz.startChallengeCommit()
+
+		return nil
+	}
+	// send up
+	return bz.SendTo(bz.Parent(), bzrReturn)
+}
+
+// computePrepareResponse wait the end of the verification and returns the
+// BizCoinResponse along with the flag:
+// true => no exception, the verification is correct
+// false => exception, the verification is NOT correct
+func (bz *BizCoin) waitResponseVerification() (*BizCoinResponse, bool) {
+	bzr := &BizCoinResponse{
+		TYPE: ROUND_PREPARE,
+	}
+	// wait the verification
+	verified := <-bz.verifyBlockChan
+	if !verified {
+		// append our exception
+		bzr.Exceptions = append(bzr.Exceptions, cosi.Exception{bz.Public(), bz.prepare.GetCommitment()})
+		bz.sendAndMeasureViewchange()
+		return bzr, false
+	}
+
+	return bzr, true
+}
+
+// verifyBlock is a simulation of a real verification block algorithm
+func verifyBlock(block *blockchain.TrBlock, lastBlock, lastKeyBlock string, done chan bool) {
+	//We measure the average block verification delays is 174ms for an average
+	//block of 500kB.
+	//To simulate the verification cost of bigger blocks we multiply 174ms
+	//times the size/500*1024
+	b, _ := json.Marshal(block)
+	s := len(b)
+	var n time.Duration
+	n = time.Duration(s / (500 * 1024))
+	time.Sleep(150 * time.Millisecond * n) //verification of 174ms per 500KB simulated
+	// verification of the header
+	verified := block.Header.Parent == lastBlock && block.Header.ParentKey == lastKeyBlock
+	verified = verified && block.Header.MerkleRoot == blockchain.HashRootTransactions(block.TransactionList)
+	verified = verified && block.HeaderHash == blockchain.HashHeader(block.Header)
+	// notify it
+	dbg.Lvl3("Verification of the block done =", verified)
+	done <- verified
+}
+
+// getblock returns the next block available from the transaction pool.
+func getBlock(transactions []blkparser.Tx, lastBlock, lastKeyBlock string) (*blockchain.TrBlock, error) {
+	if len(transactions) < 1 {
+		return nil, errors.New("no transaction available")
+	}
+
+	trlist := blockchain.NewTransactionList(transactions, len(transactions))
+	header := blockchain.NewHeader(trlist, lastBlock, lastKeyBlock)
+	trblock := blockchain.NewTrBlock(trlist, header)
+	return trblock, nil
+}
+
+// Signature will generate the final signature, the output of the BizCoin
+// protocol.
+func (bz *BizCoin) Signature() *BlockSignature {
+	return &BlockSignature{
+		Sig:        bz.commit.Signature(),
+		Block:      bz.tempBlock,
+		Exceptions: bz.tempExceptions,
+	}
+}
+
+func (bz *BizCoin) RegisterOnDone(fn func()) {
```

Comment:
It's more to have a unified way of doing things. For the rest of the protocols we always used callbacks each time the protocol has finished (the OnDone). For clarity we just use the same mechanism. The fact that it's only used in server.go to pass that to a channel does not matter to bizcoin.go. One could use bizcoin.go in other ways.
Why we use a channel after is only for the simulation, it's easier to wait on a channel that a callback.